File: native-modules/react-native-bundle-update/android/src/main/java/com/margelo/nitro/reactnativebundleupdate/BundleStorageSchemaPolicy.kt
package com.margelo.nitro.reactnativebundleupdate

internal object BundleStorageSchemaPolicy {
    const val METADATA_KEY = "storageSchemaVersion"
    const val SUPPORTED_VERSION = "mmkv-v1"
    const val LEDGER_PREFS_NAME = "onekey_native_storage_migration"
    private val ledgerKeys = listOf("app-storage-v1", "jotai-storage-v1")

    fun isCompatible(
        declaredVersion: String?,
        readMigrationLedger: (String) -> String?,
    ): Boolean {
        if (!declaredVersion.isNullOrEmpty()) {
            return declaredVersion == SUPPORTED_VERSION
        }

        // Old apps do not write this ledger and may keep using unmarked bundles.
        // Any non-empty state (including an interrupted migration or reset)
        // forbids returning to the retained AsyncStorage database.
        return ledgerKeys.none { key -> !readMigrationLedger(key).isNullOrEmpty() }
    }
}

File: native-modules/react-native-bundle-update/ios/BundleStorageSchemaPolicy.swift
enum BundleStorageSchemaPolicy {
    static let metadataKey = "storageSchemaVersion"
    static let supportedVersion = "mmkv-v1"
    static let ledgerPrefix = "onekey_native_storage_migration_"
    private static let ledgerKeys = ["app-storage-v1", "jotai-storage-v1"]

    static func isCompatible(
        declaredVersion: String?,
        readMigrationLedger: (String) -> String?
    ) -> Bool {
        if let declaredVersion, !declaredVersion.isEmpty {
            return declaredVersion == supportedVersion
        }

        // Old apps do not write this ledger and may keep using unmarked bundles.
        // Any non-empty state (including an interrupted migration or reset)
        // forbids returning to the retained AsyncStorage database.
        return !ledgerKeys.contains { key in
            !(readMigrationLedger(ledgerPrefix + key) ?? "").isEmpty
        }
    }
}
